pkg/pool-weighted/contracts/lib/CircuitBreakerLib.sol

```diff
@@ -0,0 +1,299 @@
+ * and go to zero. With no safeguards, arbitrageurs could drain the pool by selling large amounts of the
+ * token to the pool at inflated internal prices.
+ *
+ * The circuit breaker mechanism establishes a "safe trading range" for each token, expressed in terms of
+ * the BPT price. Both lower and upper bounds can be set, and if a trade would result in moving the BPT price
+ * of any token involved in the operation outside that range, the breaker is "tripped", and the operation
+ * should revert. Each token is independent, since some might have very "tight" valid trading ranges, such as
+ * stable coins, and others are more volatile.
+ *
+ * The BPT price of a token is defined as the amount of BPT that could be redeemed for a single token.
+ * For instance, in an 80/20 pool with a total supply of 1000, the 80% token accounts for 800 BPT. So each
+ * token would be worth 800 / token balance. The formula is then: total supply * token weight / token balance.
+ * (Note that this only applies *if* the pool is balanced (a condition that cannot be checked by the pool without
+ * accessing price oracles.)
+ *
+ * We need to use the BPT price as the measure to ensure we account for the change relative to the rest of
+ * the pool, which could have many other tokens. The drop detected by circuit breakers is analogous to
+ * impermanent loss: it is relative to the performance of the other tokens. If the entire market tanks and
+ * all token balances go down together, the *relative* change would be zero, and the breaker would not be
+ * triggered: even though the external price might have dropped 50 or 70%. It is only the *relative* movement
+ * compared to the rest of the pool that matters.
+ *
+ * If we have tokens A, B, and C, If A drops 20% and B and C are unchanged, that's a simple 20% drop for A.
+ * However, if A is unchanged and C increases 25%, that would also be a 20% "drop" for A 1 / 1.25 = 0.8.
+ * The breaker might register a 20% drop even if both go up - if our target token lags the market. For
+ * instance, if A goes up 60% and B and C double, 1.6 / 2 = 0.8.
+ *
+ * Since BPT prices are not intuitive - and there is a very non-linear relationship between "spot" prices and
+ * BPT prices - circuit breakers are set using simple percentages. Intuitively, a lower bound of 0.8 means the
+ * token can lose 20% of its value before triggering the circuit breaker, and an upper bound of 3.0 means it
+ * can triple before being halted. These percentages are then transformed into BPT prices for comparison to the
+ * "reference" state of the pool when the circuit breaker was set.
+ *
+ * Prices can change in two ways: arbitrage traders responding to external price movement can change the balances,
+ * or an ongoing gradual weight update (or change in pool composition) can change the weights. In order to isolate
+ * the balance changes due to price movement, the bounds are dynamic, adjusted for the current weight.
+ */
+library CircuitBreakerLib {
```

Comment:
Seems like this should expose a `hasCircuitBreakerTripped(bytes32 circuitBreakerState, uint256 currentWeightComplement, uint256 tokenBalance)` function. This would avoid us having to calculate the BPT price inside `ManagedPool`.

Reply:
Look at the next PR (#1834). You have to calculate the bptPrice anyway when setting it, since it's one of the inputs. I have a getBptPrice method (public, since it might be externally, e.g., for an aggregator predicting the behavior of the pool).

I suppose it could be pushed down to the library, but it needs so much "pool stuff" the library doesn't know about to calculate it: supply, scaling factor for the balance, etc., I thought it made more sense to be in the pool.

Reply:
I think we should definitely have a version of this function in the library for a couple reasons.
1. It doesn't really feel complete to have a circuit breaker library which doesn't have a "am I tripped" function.
2. It would be a good example of usage even if we were to have a parallel implementation somewhere else. This would be useful for testing.
3. It fits in the ethos of `ManagedPool` being glue code and we'll have all the relevant inputs calculated already so we can just feed them into this function quite naturally.